### PR TITLE
Batched GEMM Multiple D based on Universal GEMM

### PR DESCRIPTION
**Pull request by zjing14**

### CODE REVIEW

File: example/24_batched_gemm/batched_gemm_xdl_bf16_v3.cpp [hunk 1]
```diff
@@ -0,0 +1,97 @@
+#include <iostream>
```
Comment (bartekxk) [inline example/24_batched_gemm/batched_gemm_xdl_bf16_v3.cpp:3]:
```suggestion
// SPDX-License-Identifier: MIT
// Copyright (c) 2024, Advanced Micro Devices, Inc. All rights reserved.
#include <iostream>
```

File: example/24_batched_gemm/batched_gemm_xdl_fp8_rowwise_v3.cpp [hunk 2]
```diff
@@ -0,0 +1,104 @@
+#include <iostream>
```
Comment (bartekxk) [inline example/24_batched_gemm/batched_gemm_xdl_fp8_rowwise_v3.cpp:3]:
```suggestion
// SPDX-License-Identifier: MIT
// Copyright (c) 2024, Advanced Micro Devices, Inc. All rights reserved.
#include <iostream>
```

File: example/24_batched_gemm/run_batched_gemm_example.inc [hunk 3]
```diff
@@ -79,9 +79,15 @@ bool run_batched_gemm(const ProblemSize& problem_size, const ExecutionConfig& co
         f_host_tensor_descriptor(batch_count, M, N, stride_C, batch_stride_C, ELayout{}));
 #endif
 
-    std::cout << "a_g_m_k: " << a_g_m_k.mDesc << std::endl;
-    std::cout << "b_g_k_n: " << b_g_k_n.mDesc << std::endl;
-    std::cout << "e_g_m_n: " << e_g_m_n_device_result.mDesc << std::endl;
+    std::cout << "a_g_m_k: " << a_g_m_k.mDesc << ", size = "
```
Comment (bartekxk) [inline example/24_batched_gemm/run_batched_gemm_example.inc]: Can we make some constexpr for this 1024 * 1024?

File: example/24_batched_gemm/run_batched_gemm_example_rowwise.inc [hunk 4]
```diff
@@ -0,0 +1,288 @@
+#include <random>
```
Comment (bartekxk) [inline example/24_batched_gemm/run_batched_gemm_example_rowwise.inc:3]:
```suggestion
// SPDX-License-Identifier: MIT
// Copyright (c) 2024, Advanced Micro Devices, Inc. All rights reserved.
#include <random>
```

File: example/24_batched_gemm/run_batched_gemm_example_rowwise.inc [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+        f_host_tensor_descriptor(batch_count, K, N, stride_B, batch_stride_B, BLayout{}));
+    Tensor<D0DataType> d0_g_m_n(
+        f_host_tensor_descriptor(batch_count, M, N, stride_D0, batch_stride_D0, D0Layout{}));
+    Tensor<D1DataType> d1_g_m_n(
+        f_host_tensor_descriptor(batch_count, M, N, stride_D1, batch_stride_D1, D1Layout{}));
+    Tensor<EDataType> e_g_m_n_device_result(
+        f_host_tensor_descriptor(batch_count, M, N, stride_C, batch_stride_C, ELayout{}));
+
+    std::cout << "a_g_m_k: " << a_g_m_k.mDesc << ", size = "
+              << sizeof(ADataType) * a_g_m_k.mDesc.GetElementSpaceSize() / 1024 / 1024 << " MB"
```
Comment (bartekxk) [inline example/24_batched_gemm/run_batched_gemm_example_rowwise.inc]: Can we create some constexpr for these 1024 * 1024?

File: example/24_batched_gemm/run_batched_gemm_example_rowwise.inc [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+
+        Tensor<EDataType> e_g_m_n_host_result(
+            f_host_tensor_descriptor(batch_count, M, N, stride_C, batch_stride_C, ELayout{}));
+
+        auto ref_argument = ref_batched_gemm.MakeArgument(
+            a_g_m_k, b_g_k_n, c_g_m_n, a_element_op, b_element_op, PassThrough{});
+
+        ref_invoker.Run(ref_argument);
+
+        for(int b = 0; b < batch_count; ++b)
```
Comment (bartekxk) [inline example/24_batched_gemm/run_batched_gemm_example_rowwise.inc:185]: We have out_element_op argument in`ReferenceBatchedGemmInstance` can we use this one?

File: include/ck/tensor_operation/gpu/device/device_batched_gemm_multi_d.hpp [hunk 7]
```diff
@@ -53,6 +53,47 @@ struct DeviceBatchedGemmMultiD : public BaseOperator
...
+          typename DsLayout,
+          typename ELayout,
+          typename ADataType,
+          typename BDataType,
+          typename DsDataType,
+          typename EDataType,
+          typename AElementwiseOperation,
+          typename BElementwiseOperation,
+          typename CDEElementwiseOperation>
+struct DeviceBatchedGemmV2MultiD : public BaseOperator
```
Comment (bartekxk) [inline include/ck/tensor_operation/gpu/device/device_batched_gemm_multi_d.hpp:67]: What is the diffrence between this Device op and DeviceBatchedGemmMultiD?

File: include/ck/tensor_operation/gpu/device/impl/device_batched_gemm_multiple_d_xdl_cshuffle_v3.hpp [hunk 8]
```diff
@@ -0,0 +1,1012 @@
...
+    kernel_batched_gemm_xdl_cshuffle_v3_multi_d(BatchedGemmArg karg)
+{
+#if(!defined(__HIP_DEVICE_COMPILE__) || defined(__gfx9__))
+    __shared__ char p_shared[GridwiseGemm::GetSharedMemoryNumberOfByte()];
+
+    // auto splitk_batch_offset = typename GridwiseGemm::SplitKBatchOffset(karg);
+
+    // const index_t num_blocks_per_batch =
+    //__builtin_amdgcn_readfirstlane(get_grid_size() / karg.Batch);
+    // const index_t g_idx = __builtin_amdgcn_readfirstlane(get_grid_size() / num_blocks_per_batch);
```
Comment (bartekxk) [inline include/ck/tensor_operation/gpu/device/impl/device_batched_gemm_multiple_d_xdl_cshuffle_v3.hpp]:
```suggestion
```

File: include/ck/tensor_operation/gpu/device/impl/device_batched_gemm_multiple_d_xdl_cshuffle_v3.hpp [hunk 9]
```diff
@@ -0,0 +1,1012 @@
...
+          typename BatchedGemmArg,
+          bool HasMainKBlockLoop,
+          InMemoryDataOperationEnum CGlobalMemoryDataOperation,
+          index_t MinimumOccupancy = 1,
+          TailNumber TailNum       = TailNumber::Full>
+__global__ void
+#if CK_USE_LAUNCH_BOUNDS
+    __launch_bounds__(CK_MAX_THREAD_PER_BLOCK, MinimumOccupancy)
+#endif
+    // __attribute__((amdgpu_waves_per_eu(1, 1)))
```
Comment (bartekxk) [inline include/ck/tensor_operation/gpu/device/impl/device_batched_gemm_multiple_d_xdl_cshuffle_v3.hpp]:
```suggestion
```

File: include/ck/tensor_operation/gpu/device/impl/device_batched_gemm_multiple_d_xdl_cshuffle_v3.hpp [hunk 10]
```diff
@@ -0,0 +1,1012 @@
...
```
Comment (bartekxk) [inline include/ck/tensor_operation/gpu/device/impl/device_batched_gemm_multiple_d_xdl_cshuffle_v3.hpp]:
```suggestion
```

File: include/ck/tensor_operation/gpu/device/impl/device_batched_gemm_multiple_d_xdl_cshuffle_v3.hpp [hunk 11]
```diff
@@ -0,0 +1,1012 @@
...
+#if(!defined(__HIP_DEVICE_COMPILE__) || defined(__gfx9__))
+    // Pass two lds pointer is the key to tell compiler that ds_read/write
+    // operate on different lds chunk at same time without order dependecy
+    __shared__ char p_shared_0[GridwiseGemm::GetSharedMemoryNumberOfByte()];
+    __shared__ char p_shared_1[GridwiseGemm::GetSharedMemoryNumberOfByte()];
+
+    // auto splitk_batch_offset = typename GridwiseGemm::SplitKBatchOffset(karg);
+
+    // const index_t num_blocks_per_batch =
+    //__builtin_amdgcn_readfirstlane(get_grid_size() / karg.Batch);
```
Comment (bartekxk) [inline include/ck/tensor_operation/gpu/device/impl/device_batched_gemm_multiple_d_xdl_cshuffle_v3.hpp]:
```suggestion
```

File: include/ck/tensor_operation/gpu/device/impl/device_batched_gemm_multiple_d_xdl_cshuffle_v3.hpp [hunk 12]
```diff
@@ -0,0 +1,1012 @@
...
+                                     a_element_op_,
+                                     b_element_op_,
+                                     c_element_op_},
+              Batch{Batch_},
+              BatchStrideDs{BatchStrideDs_},
+              compute_ptr_offset_of_batch{BatchStrideA_, BatchStrideB_, BatchStrideE_}
+        {
+        }
+    };
+    // using Argument = typename GridwiseGemm::Argument;
```
Comment (bartekxk) [inline include/ck/tensor_operation/gpu/device/impl/device_batched_gemm_multiple_d_xdl_cshuffle_v3.hpp]:
```suggestion
```

File: library/src/tensor_operation_instance/gpu/gemm_universal_batched/device_batched_gemm_xdl_universal_bf16_bf16_bf16/device_batched_gemm_xdl_universal_bf16_bf16_bf16_mk_nk_mn_comp_default_instance.cpp [hunk 13]
```diff
@@ -0,0 +1,32 @@
+// SPDX-License-Identifier: MIT
+// Copyright (c) 2018-2023, Advanced Micro Devices, Inc. All rights reserved.
```
Comment (bartekxk) [inline library/src/tensor_operation_instance/gpu/gemm_universal_batched/device_batched_gemm_xdl_universal_bf16_bf16_bf16/device_batched_gemm_xdl_universal_bf16_bf16_bf16_mk_nk_mn_comp_default_instance.cpp]:
```suggestion
// Copyright (c) 2024, Advanced Micro Devices, Inc. All rights reserved.
```

File: library/src/tensor_operation_instance/gpu/gemm_universal_batched/device_batched_gemm_xdl_universal_bf16_bf16_bf16/device_batched_gemm_xdl_universal_bf16_bf16_bf16_mk_nk_mn.hpp [hunk 14]
```diff
@@ -0,0 +1,95 @@
+// SPDX-License-Identifier: MIT
+// Copyright (c) 2018-2023, Advanced Micro Devices, Inc. All rights reserved.
```
Comment (bartekxk) [inline library/src/tensor_operation_instance/gpu/gemm_universal_batched/device_batched_gemm_xdl_universal_bf16_bf16_bf16/device_batched_gemm_xdl_universal_bf16_bf16_bf16_mk_nk_mn.hpp]:
```suggestion
// Copyright (c) 2024, Advanced Micro Devices, Inc. All rights reserved.
```

File: library/src/tensor_operation_instance/gpu/gemm_universal_batched/device_batched_gemm_xdl_universal_bf16_bf16_bf16/device_batched_gemm_xdl_universal_bf16_bf16_bf16_mk_nk_mn_mem_v1_default_instance.cpp [hunk 15]
```diff
@@ -0,0 +1,33 @@
+// SPDX-License-Identifier: MIT
+// Copyright (c) 2018-2023, Advanced Micro Devices, Inc. All rights reserved.
```
Comment (bartekxk) [inline library/src/tensor_operation_instance/gpu/gemm_universal_batched/device_batched_gemm_xdl_universal_bf16_bf16_bf16/device_batched_gemm_xdl_universal_bf16_bf16_bf16_mk_nk_mn_mem_v1_default_instance.cpp]:
```suggestion
// Copyright (c) 2024, Advanced Micro Devices, Inc. All rights reserved.
```

File: library/src/tensor_operation_instance/gpu/gemm_universal_batched/device_batched_gemm_xdl_universal_bf16_bf16_bf16/device_batched_gemm_xdl_universal_bf16_bf16_bf16_mk_nk_mn_mem_v2_default_instance.cpp [hunk 16]
```diff
@@ -0,0 +1,33 @@
+// SPDX-License-Identifier: MIT
+// Copyright (c) 2018-2023, Advanced Micro Devices, Inc. All rights reserved.
```
Comment (bartekxk) [inline library/src/tensor_operation_instance/gpu/gemm_universal_batched/device_batched_gemm_xdl_universal_bf16_bf16_bf16/device_batched_gemm_xdl_universal_bf16_bf16_bf16_mk_nk_mn_mem_v2_default_instance.cpp]:
```suggestion
// Copyright (c) 2024, Advanced Micro Devices, Inc. All rights reserved.
```

File: profiler/include/profiler/profile_gemm_universal_batched_impl.hpp [hunk 17]
```diff
@@ -0,0 +1,280 @@
+// SPDX-License-Identifier: MIT
+// Copyright (c) 2018-2023, Advanced Micro Devices, Inc. All rights reserved.
```
Comment (bartekxk) [inline profiler/include/profiler/profile_gemm_universal_batched_impl.hpp]:
```suggestion
// Copyright (c) 2024, Advanced Micro Devices, Inc. All rights reserved.
```

File: profiler/src/profile_gemm_universal_batched.cpp [hunk 18]
```diff
@@ -0,0 +1,183 @@
+// SPDX-License-Identifier: MIT
+// Copyright (c) 2018-2023, Advanced Micro Devices, Inc. All rights reserved.
```
Comment (bartekxk) [inline profiler/src/profile_gemm_universal_batched.cpp]:
```suggestion
// Copyright (c) 2024, Advanced Micro Devices, Inc. All rights reserved.
```

**Review by bartekxk — COMMENTED** [review]
Thank you @zjing14, added a few comments, most of them are minor

File: example/24_batched_gemm/run_batched_gemm_example.inc (hunk 3 above)
Comment (zjing14) [reply example/24_batched_gemm/run_batched_gemm_example.inc]: These are debug info. Removed it.

File: example/24_batched_gemm/run_batched_gemm_example_rowwise.inc (hunk 5 above)
Comment (zjing14) [reply example/24_batched_gemm/run_batched_gemm_example_rowwise.inc]: These are debug info. Removed it.

**Comment by zjing14** [comment]
@bartekxk Please review the PR again.

File: include/ck/tensor_operation/gpu/device/device_batched_gemm_multi_d.hpp (hunk 7 above)
Comment (zjing14) [reply include/ck/tensor_operation/gpu/device/device_batched_gemm_multi_d.hpp:67]: It is based on a different gridwise_gemm, which has different template parameters.

File: example/24_batched_gemm/run_batched_gemm_example_rowwise.inc (hunk 6 above)
Comment (zjing14) [reply example/24_batched_gemm/run_batched_gemm_example_rowwise.inc:185]: `ReferenceBatchedGemmInstance`'s output_elementwise does not support multiple Ds.

File: include/ck/tensor_operation/gpu/device/device_batched_gemm_multi_d.hpp (hunk 7 above)
Comment (bartekxk) [reply include/ck/tensor_operation/gpu/device/device_batched_gemm_multi_d.hpp:67]: Sure but why we nee diffrenet "DeviceBatchedGemm" for them? Can  DeviceBatchedGemmMultiD be base class for DeviceBatchedGemmMultiD_Xdl_CShuffle_V3?

**Review by bartekxk — CHANGES_REQUESTED** [review]
Generally LGTM, but can DeviceBatchedGemmMultiD be base class for DeviceBatchedGemmMultiD_Xdl_CShuffle_V3? We can support universal batched gemm and "old" batched gemm using one DeviceBatchedGemmMultiD.

**Comment by zjing14** [comment]
> Generally LGTM, but can DeviceBatchedGemmMultiD be base class for DeviceBatchedGemmMultiD_Xdl_CShuffle_V3? We can support universal batched gemm and "old" batched gemm using one DeviceBatchedGemmMultiD.

We want to differentiate V3 as a new pipeline.

**Comment by bartekxk** [comment]
> > Generally LGTM, but can DeviceBatchedGemmMultiD be base class for DeviceBatchedGemmMultiD_Xdl_CShuffle_V3? We can support universal batched gemm and "old" batched gemm using one DeviceBatchedGemmMultiD.
> 
> We want to differentiate V3 as a new pipeline.

Sure, but can differentate it by DeviceBatchedGemmMultiD_Xdl_CShuffle_V3? Do we need separate instance factory and seperate BaseOp for them?

**Comment by zjing14** [comment]
> > > Generally LGTM, but can DeviceBatchedGemmMultiD be base class for DeviceBatchedGemmMultiD_Xdl_CShuffle_V3? We can support universal batched gemm and "old" batched gemm using one DeviceBatchedGemmMultiD.
> > 
> > 
> > We want to differentiate V3 as a new pipeline.
> 
> Sure, but can differentate it by DeviceBatchedGemmMultiD_Xdl_CShuffle_V3? Do we need separate instance factory and seperate BaseOp for them?

It will have too much instances from old BMM kernel, which has bad performance.

**Review by bartekxk — DISMISSED** [review]

**Comment by bartekxk** [comment]
> > > > Generally LGTM, but can DeviceBatchedGemmMultiD be base class for DeviceBatchedGemmMultiD_Xdl_CShuffle_V3? We can support universal batched gemm and "old" batched gemm using one DeviceBatchedGemmMultiD.
> > > 
> > > 
> > > We want to differentiate V3 as a new pipeline.
> > 
> > 
> > Sure, but can differentate it by DeviceBatchedGemmMultiD_Xdl_CShuffle_V3? Do we need separate instance factory and seperate BaseOp for them?
> 
> It will have too much instances from old BMM kernel, which has bad performance.

Ok, so approved, thanks, I will run CI

**Comment by bartekxk** [comment]
I ran CI under this PR #1655

**Comment by bartekxk** [comment]
Hi @zjing14 , build of profile_gemm_universal_batched failed. Could you fix that under #1655 ?

**Comment by zjing14** [comment]
> profile_gemm_universal_batched

Failure on which machine? I can build it without any issue on gfx942.

**Comment by bartekxk** [comment]
> > profile_gemm_universal_batched
> 
> Failure on which machine? I can build it without any issue on gfx942.

gfx90a

**Comment by zjing14** [comment]
@bartekxk Please give a try.

**Comment by bartekxk** [comment]
@zjing14  thanks, I applied and ran CI

**Comment by bartekxk** [comment]
Merged under #1655 , thanks

**Pull request closed without merging**